### PR TITLE
[PropertyInfo] Fix failure over array<string,mixed> PhpDoc type

### PR DESCRIPTION
**Pull request by romaricdrigon**
| Q             | A
| ------------- | ---
| Branch?       | master (issue introduced in master)
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no 
| Tickets       | /
| License       | MIT
| Doc PR        | /

**Background**

Since a few days ago ([this commit](https://github.com/symfony/symfony/commit/80e5a7fbd80ef04fced05bb3d4bcb8816fa897be)), PropertyInfo 5.2 component supports array PhpDoc types, such as `array<string,string>`.  
While testing SF 5.2 over my project, it was failing over:
```php
class SomeDoctrineEntityProxy implements \Doctrine\ORM\Proxy\Proxy
{
    /**
     * @var array<string, mixed> default values of properties to be lazy loaded, with keys being the property names
     *
     * @see \Doctrine\Common\Proxy\Proxy::__getLazyProperties
     */
    public static $lazyPropertiesDefaults = array();
}
```

Upon investigation, it appears `array<string,mixed>` is causing the issue: `mixed` type is not handled.

**Expected behavior**

PropertyInfo component should not guess anything for `mixed` type (return `NULL`), but it should not throw any exception.

**Comment about my fix**

I added a test case (`arrayOfMixed`) you can use to reproduce the issue.
Since array PhpDoc types were not tested at all, I also added a test case over `array<string,string>`.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Can you have a look at deps=low failures please?

**Comment by romaricdrigon** [comment]
> Can you have a look at [deps=low failures](https://travis-ci.org/github/symfony/symfony/jobs/732247821#L6219), please?

Sure, I'm ont it.

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php [hunk 1]
```diff
@@ -352,3 +368,15 @@ public function setOmittedType(array $omittedTagType)
     {
     }
 }
+
+class UnionTypeDocBlock
+{
+    /**
+     * Here we use an union type, as commonly used in PhpDoc, to make sure it is ignored and does not trigegr an exception.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php:375]: typo: s/trigegr/trigger

**Review by nicolas-grekas — APPROVED** [review]

**Comment by romaricdrigon** [comment]
So, to sum up, the issue is coming from old PhpDocExtractor versions not supporting `array<x>` notations. I believe support was introduced in version 4+ of `phpdocumentor/reflection-docblock` (for instance [this issue](https://github.com/phpDocumentor/TypeResolver/issues/89) has a few details...).
With `deps=high`, we are using version 5, it works. With `deps=low`, we are using version 3, which won't guess any type (so return NULL, and test case won't work).

 `PhpDocExtractor` code is quite intricated with it, ut is not quite obvious something an be mocked / worked around. There was already a method to detect if we are using PhpDocumentor5, so using it, test case can adapt and work with both 3 and 5. That's the best option I could think of.
Another option would be not to test that type is detected, since we can't control what PhpDocumentor is outputting. So I could remove [those lines](https://github.com/symfony/symfony/pull/38385/files#diff-0ecd6142af9284668fce7690f37fdd08R122-R124), the test would only check no exception is thrown.

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php (hunk 1 above)
Comment (romaricdrigon) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php:375]: Good spot! And actually that code is useless now, I'm removing it.

File: src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php [hunk 2]
```diff
@@ -122,7 +122,7 @@ private function createType(DocType $type, bool $nullable, string $docType = nul
             $collectionKeyType = $this->getTypes($type->getKeyType())[0];
 
             $collectionValueTypes = $this->getTypes($type->getValueType());
-            if (\count($collectionValueTypes) > 1) {
+            if (0 === \count($collectionValueTypes) || \count($collectionValueTypes) > 1) {
```
Comment (dunglas) [inline src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php:125]: `!== 1` ?

**Review by dunglas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @romaricdrigon.

**Pull request merged**